File: internal/perivale/perivale.go
package perivale

import (
	"container/list"
	"errors"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"strings"
	"sync/atomic"
	"time"

	"github.com/nosborn/federation-1999/internal/ioctl"
	"github.com/nosborn/federation-1999/internal/link"
	"github.com/nosborn/federation-1999/internal/model"
	"github.com/nosborn/federation-1999/pkg/ibgames"
	"github.com/nosborn/federation-1999/pkg/ibgames/rules"
	"golang.org/x/sys/unix"
	"golang.org/x/term"
)

type InState int

const (
	inReady InState = iota
	inAckWait
	inThrottle
)

var (
	winch int32 = 1 // Fake it the first time through

	rows int //nolint:unused
	cols int

	inState      = inReady
	unthrottleAt time.Time

	iq = list.New()
	oq = list.New()

	inputLine    = make([]byte, 256)
	inputLineLen = 0
)

func Perivale(uid ibgames.AccountID, remoteAddr string, robobod int) {
	// //
	// term := os.Getenv("TERM")
	// if term != "" {
	// 	//
	// }

	// Check for lock out!
	if rules.IsLockedOut(uid) {
		lockout()
	}

	var remoteHostname string
	if names, err := net.LookupAddr(remoteAddr); err != nil {
		remoteHostname = names[0]
	} else {
		remoteHostname = remoteAddr // Use the dotted quad instead
	}

	login := fmt.Sprintf("%d %s %s %d\n", uid, os.Args[2], remoteHostname, robobod)

	unixAddr, err := net.ResolveUnixAddr("unix", link.SocketPath)
	if err != nil {
		log.Fatalf("socket: %v", err)
	}
	unixConn, err := net.DialUnix("unix", nil, unixAddr)
	if err != nil {
		unavailable()
	}
	defer unixConn.Close()

	_, err = unixConn.Write([]byte(login))
	if err != nil {
		unavailable()
	}

	//
	stdinFd := int(os.Stdin.Fd())
	if err = unix.SetNonblock(stdinFd, true); err != nil {
		log.Fatalf("unix.SetNonblock: %v", err)
	}
	stdoutFd := int(os.Stdout.Fd())
	if err = unix.SetNonblock(stdoutFd, true); err != nil {
		log.Fatalf("unix.SetNonblock: %v", err)
	}

	f, _ := unixConn.File()
	connFd := int(f.Fd())

	oldTermios, err := unix.IoctlGetTermios(int(os.Stdin.Fd()), ioctl.GetTerminalAttrs)
	if err != nil {
		log.Printf("IoctlGetTermios: %v", err)
	} else {
		defer func() {
			unix.IoctlSetTermios(int(os.Stdin.Fd()), ioctl.SetTerminalAttrsNow, oldTermios)
		}()

		newTermios := *oldTermios
		newTermios.Iflag |= unix.ICRNL | unix.IGNBRK | unix.ISTRIP
		newTermios.Lflag |= unix.ICANON
		newTermios.Lflag &^= unix.ECHO | unix.ISIG
		newTermios.Cc[unix.VINTR] = 0
		newTermios.Cc[unix.VQUIT] = 0
		newTermios.Cc[unix.VKILL] = 0

		if err := unix.IoctlSetTermios(int(os.Stdin.Fd()), ioctl.SetTerminalAttrsNow, &newTermios); err != nil {
			log.Printf("IoctlSetTermios: %v", err)
		}
	}

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, unix.SIGWINCH)
	go func() {
		for range sigCh {
			sigWinch()
		}
	}()

	selector, err := NewSelector()
	if err != nil {
		log.Fatalf("NewSelector: %v", err)
	}
	defer selector.Close()

	selector.AddRead(stdinFd)
	selector.AddRead(connFd)

	for {
		selector.UpdateWrite(stdoutFd, oq.Len() > 0)

		timeout := -1 * time.Second // Negative timeout blocks indefinitely
		if inState == inThrottle {
			now := time.Now()
			if now.After(unthrottleAt) {
				inState = inReady
				timeout = 0 // Poll, don't block
			} else {
				timeout = unthrottleAt.Sub(now)
			}
		}

		events, err := selector.Wait(timeout)
		if err != nil {
			log.Fatalf("wait: %v", err)
		}

		if atomic.CompareAndSwapInt32(&winch, 1, 0) {
			var err error
			cols, rows, err = term.GetSize(int(os.Stdin.Fd()))
			if err != nil {
				log.Fatal("TIOCGWINSZ")
			}
		}

		if len(events) == 0 && inState == inThrottle {
			inState = inReady
		}

		for _, event := range events {
			switch event.Fd {
			case connFd:
				if event.IsEOF {
					flushAllOutput()
					time.Sleep(1 * time.Second)
					os.Exit(0)
				}
				buf := make([]byte, 4096)
				n, err := unixConn.Read(buf)
				if err != nil {
					if opErr, ok := err.(*net.OpError); ok && opErr.Err == unix.EAGAIN {
						break // Nothing to read
					}
					log.Printf("read: %v", err)
					flushAllOutput()
					os.Exit(1)
				}

				if n == 0 { // EOF from game server
					flushAllOutput()
					os.Exit(0)
				}

				if n > 0 {
					output(buf[:n])
				}
			case stdoutFd:
				flushOutput()
			case stdinFd:
				readInput()
			}
		}

		if iq.Len() > 0 && inState == inReady {
			element := iq.Front()
			line := element.Value.([]byte)
			_, err := unixConn.Write(line)
			if err != nil {
				if opErr, ok := err.(*net.OpError); ok && opErr.Err == unix.EAGAIN {
					return
				}
				log.Fatalf("write: %v", err)
			}
			iq.Remove(element)
			inState = inAckWait
		}
	}
}

func flushAllOutput() {
	unix.SetNonblock(int(os.Stdout.Fd()), false)
	for oq.Len() > 0 {
		flushOutput()
	}
}

func flushOutput() {
	for oq.Len() > 0 {
		element := oq.Front()
		data := element.Value.([]byte)

		n, err := os.Stdout.Write(data)
		if err != nil {
			if errors.Is(err, unix.EAGAIN) {
				if n > 0 {
					element.Value = data[n:]
				}
				return
			}
			log.Fatalf("write to stdout failed: %v", err)
		}

		if n < len(data) {
			element.Value = data[n:]
			return
		}
		oq.Remove(element)
	}
}

func lockout() {
	fmt.Println("")
	fmt.Println("Your account has been locked out of the Federation game. If you")
	fmt.Println("are unsure about why you have been locked out of the game, please")
	fmt.Println("read the rules at <URL:https://federation-1999.fly.dev/ibinfo/t&c.html>")
	fmt.Printf("or send e-mail to %s.\n\n", rules.INFO_EMAIL)
	time.Sleep(5 * time.Second)
	os.Exit(0)
}

type State int

const (
	tsData State = iota
	tsSpyDepth
	tsTrace
	tsEscape
)

var (
	state        State  = tsData
	trace        bool   = false
	blankPending bool   = true
	column       int    = 1
	allWS        bool   = false
	eatBlanks    bool   = false
	lastBlank    bool   = false
	lastSpyDepth int    = 0
	prefix       string = ""
	spyDepth     int    = 0
	continuation bool   = false
)

func output(buf []byte) {
	var output, line strings.Builder

	if trace {
		fmt.Fprintf(&output, "[LEN=%d]", len(buf))
	}

	i := 0
	for i < len(buf) {
		ch := buf[i]
		i++

		switch state {
		case tsData:
			switch ch {
			case model.DLE:
				state = tsEscape
			case '\a':
				line.WriteByte(ch)
			case '\n':
				if allWS {
					if !lastBlank {
						lastBlank = true
					}
					line.Reset()
					blankPending = true
				} else {
					lastBlank = false
				}

				if line.Len() > 0 || continuation {
					trimmedLine := strings.TrimRight(line.String(), " ")
					if blankPending {
						if trace {
							output.WriteString("[PEND]")
						}
						if spyDepth == lastSpyDepth {
							output.WriteString(prefix)
						}
						output.WriteByte('\n')
						blankPending = false
					}
					if spyDepth > 0 {
						output.WriteString(prefix)
						lastSpyDepth = spyDepth
					}
					output.WriteString(trimmedLine)
					if trace {
						output.WriteString("[PARA]")
					}
					output.WriteByte('\n')
					line.Reset()
					continuation = false
				}
				column = 1
				allWS = true
				eatBlanks = false
			case '\r':
				continue
			case ' ':
				if eatBlanks {
					continue
				}
				fallthrough
			default:
				allWS = allWS && (ch == ' ')
				if eatBlanks {
					if ch == '/' || ch == '>' {
						continue
					}
					eatBlanks = false
				}

				if ch < 32 || ch > 126 {
					line.WriteByte('^')
					line.WriteByte(ch ^ 0x40)
				} else {
					line.WriteByte(ch)
				}

				column++
				if cols > 0 && column >= cols-len(prefix) {
					lineStr := line.String()
					lastSpace := strings.LastIndex(lineStr, " ")
					if lastSpace != -1 {
						chop := len(lineStr) - lastSpace - 1
						if chop > 0 {
							i -= chop
							line.Reset()
							line.WriteString(lineStr[:lastSpace+1])
						}
					}
					trimmedLine := strings.TrimRight(line.String(), " ")
					line.Reset()
					line.WriteString(trimmedLine)

					if blankPending {
						if trace {
							output.WriteString("[PEND]")
						}
						if spyDepth == lastSpyDepth {
							output.WriteString(prefix)
						}
						output.WriteByte('\n')
						blankPending = false
					}
					if spyDepth > 0 {
						output.WriteString(prefix)
						lastSpyDepth = spyDepth
					}
					output.WriteString(line.String())
					if trace {
						output.WriteString("[WRAP]")
					}
					output.WriteByte('\n')
					line.Reset()
					continuation = false
					column = 1
					allWS = true
					eatBlanks = true
				}
			}
		case tsEscape:
			switch ch {
			case model.LeAck:
				if spyDepth == 0 {
					if trace {
						output.WriteString("[ACK]")
					}
					if inState == inAckWait {
						unthrottleAt = time.Now().Add(500 * time.Millisecond)
						inState = inThrottle
					}
				}
				state = tsData
			case model.LeSpy:
				state = tsSpyDepth
			case model.LeTrace:
				state = tsTrace
			default:
				state = tsData
			}
		case tsSpyDepth:
			if trace {
				fmt.Fprintf(&output, "[S=%d]", ch)
			}
			prefix = ""
			lastSpyDepth = spyDepth
			spyDepth = int(ch)
			if spyDepth > 0 {
				prefix = strings.Repeat("/", spyDepth) + " "
			}
			state = tsData
		case tsTrace:
			if spyDepth == 0 {
				wasTrace := trace
				trace = (ch != '-')
				if trace || wasTrace {
					fmt.Fprintf(&output, "[T=%t]", trace)
				}
			}
			state = tsData
		}
	}

	if line.Len() > 0 {
		if blankPending {
			if spyDepth == lastSpyDepth {
				output.WriteString(prefix)
			}
			output.WriteByte('\n')
			blankPending = false
		}
		if spyDepth > 0 {
			output.WriteString(prefix)
			lastSpyDepth = spyDepth
		}
		output.WriteString(line.String())
		continuation = true
	}

	if output.Len() > 0 {
		writeOutput([]byte(output.String()))
	}
}

func readInput() {
	buf := make([]byte, 4096)
	n, err := os.Stdin.Read(buf)
	if err != nil {
		if err == unix.EAGAIN {
			return
		}
		log.Printf("read from stdin failed: %v", err)
		flushAllOutput()
		os.Exit(1)
	}

	if n == 0 { // EOF
		flushAllOutput()
		os.Exit(0)
	}

	for i := range n {
		ch := buf[i]
		switch ch {
		case '\n':
			for inputLineLen > 0 && inputLine[inputLineLen-1] == ' ' {
				inputLineLen--
			}
			inputLine[inputLineLen] = '\n'
			inputLineLen++
			line := make([]byte, inputLineLen)
			copy(line, inputLine[:inputLineLen])
			iq.PushBack(line)
			inputLineLen = 0
		default:
			if inputLineLen < len(inputLine) {
				inputLine[inputLineLen] = ch
				inputLineLen++
			}
		}
	}
}

func sigWinch() {
	atomic.StoreInt32(&winch, 1)
}

func unavailable() {
	message := "\n\nFederation is temporarily unavailable. Please try again later.\n\n"

	time.Sleep(2 * time.Second)

	fmt.Fprintln(os.Stdout, message)
	// fflush(stdout);

	time.Sleep(3 * time.Second)
	os.Exit(0)
}

func writeOutput(buf []byte) {
	oq.PushBack(buf)
}

File: cmd/perivale/perivale.cc
/******************************************************************************

  Copyright (c) 1997-1999 Alan Lenton & Interactive Broadcasting Ltd.
  All Rights Reserved.

  No part of this software may be reproduced, transmitted, transcribed, stored
  in a retrieval system, or translated into any human or computer language, in
  any form or by any means, electronic, mechanical, magnetic, optical, manual
  or otherwise, without the express written permission of the copyright holder.

  $Id: perivale.cc,v 1.8 1999/04/24 21:03:39 nick Exp $

******************************************************************************/

#include <algorithm>
#include <string>

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/un.h>
#include <netdb.h>
#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <termios.h>
#include <unistd.h>

#include <ibgames/goodies.h>
#include <ibgames/rules.h>
#include <ibgames.h>

#include <link.hh>

static void error(const char *);
static void fatal(const char *) __attribute__((noreturn));
static void flushAllOutput();
static void flushOutput();
static void lockout() __attribute__((noreturn));
static void output(const unsigned char *, size_t);
static void readInput(int);
static void sigWinch(int);
static void unavailable() __attribute__((noreturn));
static void warning(const char *);
static void writeOutput(const char *, size_t);

typedef enum {
  inReady,
  inAckWait,
  inThrottle
} in_state_t;

static sig_atomic_t winch = true; // Fake it the first time through

static unsigned rows = 0;
static unsigned cols = 0;

static in_state_t inState = inReady;
static struct timeval unthrottle = { 0, 0 };

typedef struct iblock {
  struct iblock *next;
  struct iblock *prev;
  size_t len;
  char data[0];
} iblock_t;

typedef struct oblock {
  struct oblock *next;
  struct oblock *prev;
  size_t sent;
  size_t len;
  char data[0];
} oblock_t;

static iblock_t iqHead = { NULL, NULL, 0, {} };
static iblock_t iqTail = { NULL, NULL, 0, {} };
static oblock_t oqHead = { NULL, NULL, 0, 0, {} };
static oblock_t oqTail = { NULL, NULL, 0, 0, {} };

int
main(int argc, char *argv[]) {
  fflush(stdin);
  fflush(stdout);

  if (argc != 4) {
    fprintf(stderr, "usage: %s id address robobod\n", argv[0]);
    exit(EXIT_FAILURE);
  }

  //

  iqHead.next = &iqTail;
  iqTail.prev = &iqHead;
  oqHead.next = &oqTail;
  oqTail.prev = &oqHead;

  //

  const char *term = getenv("TERM");

  if (term != NULL) {
    ;
  }

  // Check for a valid UID.

  const account_id_t uid = atoi(argv[1]);

  if (uid < MIN_ACCOUNT_ID || uid >= MAX_ACCOUNT_ID) {
    fprintf(stderr, "%s: Bad UID\n", argv[0]);
    exit(EXIT_FAILURE);
  }

  // Check for lock out!

  if (isLockedOut(uid)) {
    lockout();
  }

  // Set up the login string in advance; the TP task is going to block between
  // our connect and send, so keep the window as small as possible.

  struct in_addr remote_addr;
  remote_addr.s_addr = inet_addr(argv[2]);

  const struct hostent *remote_host =
    gethostbyaddr(reinterpret_cast<char *>(&remote_addr), sizeof(remote_addr), AF_INET);

  const char *remote_hostname;

  if (remote_host == NULL) {
    remote_hostname = argv[2]; // Use the dotted quad instead
  } else {
    remote_hostname = remote_host->h_name;

    while (strlen(remote_hostname) > LINK_HOSTNAME_SIZE && *remote_hostname != '.') {
      remote_hostname++;
    }
  }

  char login[1024];
  const int len = snprintf(login, sizeof(login), LINK_FORMAT, uid, argv[2], remote_hostname, atoi(argv[3]));
  assert(len <= static_cast<int>(sizeof(login)));

  //

  struct sockaddr_un addr;

  memset(&addr, '\0', sizeof(addr));
  addr.sun_family = AF_UNIX;
  sprintf(addr.sun_path, LINK_SOCKET, IB_homeDir());

  const int sd = socket(addr.sun_family, SOCK_STREAM, 0);

  if (sd < 0) {
    fatal("socket");
  }

  if (connect(sd, reinterpret_cast<struct sockaddr *>(&addr), sizeof(addr)) < 0) {
    unavailable();
  }

  if (send(sd, login, len, 0) != len) {
    unavailable();
  }

  //

  if (IB_nonBlocking(STDIN_FILENO) == -1) {
    fatal("IB_nonBlocking");
  }

  if (IB_nonBlocking(STDOUT_FILENO) == -1) {
    fatal("IB_nonBlocking");
  }

  // Let's try this without input echo and see if there are any complaints.

  struct termios termios;

  if (tcgetattr(STDIN_FILENO, &termios) == -1) {
    warning("tcgetattr");
  } else {
    termios.c_iflag |= ICRNL;
    termios.c_iflag |= IGNBRK;
    termios.c_iflag |= ISTRIP;

    termios.c_lflag |= ICANON;
    termios.c_lflag &= ~ECHO;
    termios.c_lflag &= ~ISIG;

    termios.c_cc[VINTR] = _POSIX_VDISABLE;
    termios.c_cc[VQUIT] = _POSIX_VDISABLE;
    termios.c_cc[VKILL] = _POSIX_VDISABLE;

    if (tcsetattr(STDIN_FILENO, TCSANOW, &termios) == -1) {
      warning("tcsetattr");
    }
  }

  //

  signal(SIGINT, SIG_IGN);
  signal(SIGPIPE, SIG_IGN);
  signal(SIGQUIT, SIG_IGN);
  signal(SIGTSTP, SIG_IGN);
  signal(SIGWINCH, sigWinch);

  //

  for (;;) {
    fd_set readfds;
    FD_ZERO(&readfds);

    FD_SET(STDIN_FILENO, &readfds);
    FD_SET(sd, &readfds);

    fd_set writefds;
    FD_ZERO(&writefds);

    if (oqHead.next != &oqTail) {
      FD_SET(STDOUT_FILENO, &writefds);
    }

    struct timeval *timeout = NULL;
    struct timeval wakeup;

    if (inState == inThrottle) {
      struct timeval now;
      gettimeofday(&now, NULL);

      if (timercmp(&now, &unthrottle, >)) {
        inState = inReady;
      } else {
        wakeup.tv_sec = unthrottle.tv_sec - now.tv_sec;
        wakeup.tv_usec = unthrottle.tv_usec - now.tv_usec;
        if (wakeup.tv_usec < 0) {
          wakeup.tv_sec -= 1;
          wakeup.tv_usec += 1000000;
        }
        timeout = &wakeup;
      }
    }

    const int nfds = select(sd + 1, &readfds, &writefds, NULL, timeout);

    if (nfds == -1 && errno != EINTR) {
      fatal("select");
    }

    if (winch) {
      struct winsize size;

      if (ioctl(STDIN_FILENO, TIOCGWINSZ, &size) < 0) {
        fatal("TIOCGWINSZ");
      }

      rows = size.ws_row;
      cols = size.ws_col;

      winch = false;
    }

    if (nfds == -1) {
      continue;
    }

    if (nfds == 0 && inState == inThrottle) {
      inState = inReady;
    }

    if (nfds > 0) {
      if (FD_ISSET(sd, &readfds)) {
        unsigned char buf[32768];
        const ssize_t nbytes = read(sd, buf, sizeof(buf));

        if (nbytes <= 0) {
          flushAllOutput();
          if (nbytes == -1) {
            fatal("read");
          }
          sleep(1);
          exit(EXIT_SUCCESS);
        }

        output(buf, nbytes);
      }

      if (FD_ISSET(STDOUT_FILENO, &writefds)) {
        flushOutput();
      }

      if (FD_ISSET(STDIN_FILENO, &readfds)) {
        readInput(sd);
      }
    }

    if (iqHead.next != &iqTail && inState == inReady) {
      iblock_t *b = iqHead.next;

      if (write(sd, b->data, b->len) != static_cast<ssize_t>(b->len)) {
        fatal("write");
      }

      iqHead.next = b->next;
      b->next->prev = &iqHead;
      free(b);

      inState = inAckWait;
    }
  }
}

static void
error(const char *s) {
  const int errnum = errno;

  perror(s);
  syslog(LOG_ERR, "%s: %s", s, strerror(errnum));
}

static void
fatal(const char *s) {
  error(s);
  exit(EXIT_FAILURE);
}

static void
flushAllOutput() {
  while (oqHead.next != &oqTail) {
    flushOutput();
    sleep(1);
  }
}

static void
flushOutput() {
  assert(oqHead.next != &oqTail);

  do {
    oblock_t *b = oqHead.next;
    ssize_t sent = write(STDOUT_FILENO, b->data + b->sent, b->len);

    if (sent == -1) {
      if (errno != EAGAIN) {
        fatal("write");
      }
      sent = 0;
    }

    b->sent += sent;
    b->len -= sent;

    if (b->len > 0) {
      break;
    }

    oqHead.next = b->next;
    b->next->prev = &oqHead;
    free(b);
  } while (oqHead.next != &oqTail);
}

static void
lockout() {
  puts("");
  puts("Your account has been locked out of the Federation game. If you");
  puts("are unsure about why you have been locked out of the game, please");
  puts("read the rules at <URL:http://www.ibgames.net/ibinfo/t&c.html> or");
  printf("send e-mail to %s.\n\n", RULES_INFO_EMAIL);
  fflush(stdout);

  sleep(5);

  exit(EXIT_SUCCESS);
}

// codeql[c-cpp/poorly-documented-function] fedtpd output processing state machine - complexity is inherent
static void
output(const unsigned char *buf, size_t len) {
  typedef enum {
    tsData,
    tsSpyDepth,
    tsTrace,
    tsEscape
  } state_t;

  static state_t state = tsData;
  static bool trace = false;
  static bool blankPending = true;
  static size_t column = 1;
  static bool allWS = false;
  static bool eatBlanks = false;
  static bool lastBlank = false;
  static int lastSpyDepth = 0;
  static std::string prefix = "";
  static int spyDepth = 0;
  static bool continuation = false;

  assert(buf != NULL);

  std::string output;
  std::string line;
  char tbuf[BUFSIZ];
  bool wasTrace;

  if (trace) {
    sprintf(tbuf, "[LEN=%zu]", len);
    output += tbuf;
  }

  size_t i = 0;

  while (i < len) {
    unsigned char ch = buf[i++];

    // codeql[c-cpp/long-switch] fedtpd output state machine requires detailed case handling
    switch (state) {
      case tsData:
        // codeql[c-cpp/long-switch] DLE escape sequence handling
        switch (ch) {
          case DLE:
            state = tsEscape;
            break;

          case '\a':
            line += ch;
            break;

          case '\n':
            if (allWS) {
              if (!lastBlank) {
                lastBlank = true;
              }
              line.erase();
              blankPending = true;
            } else {
              lastBlank = false;
            }

            if (!line.empty() || continuation) {
              size_t rpos = line.find_last_not_of(' ');
              if (rpos != std::string::npos) {
                size_t chop = (line.size() - rpos) - 1;
                if (chop > 0) {
                  line.erase(rpos + 1, chop);
                }
              }
              if (blankPending) {
                if (trace) {
                  output += "[PEND]";
                }
                if (spyDepth == lastSpyDepth) {
                  output += prefix;
                }
                output += '\n';
                blankPending = false;
              }
              if (spyDepth > 0) {
                output += prefix;
                lastSpyDepth = spyDepth;
              }
              output += line;
              if (trace) {
                output += "[PARA]";
              }
              output += '\n';
              line.erase();
              continuation = false;
            }

            column = 1;
            allWS = true;
            eatBlanks = false;

            break;

          case '\r': // There shouldn't be any!
            continue;

          case ' ':
            if (eatBlanks) {
              continue;
            }
            // FALL THROUGH

          default:
            allWS &= (ch == ' ');

            if (eatBlanks) {
              if (ch == '/' || ch == '>') {
                continue;
              }
              eatBlanks = false;
            }

            if (!isascii(ch) || !isprint(ch)) {
              line += '^';
              line += ch ^ 0100;
            } else {
              line += ch;
            }

            if (column++ == cols - prefix.length()) {
              size_t rpos = line.find_last_of(' ');
              if (rpos != std::string::npos) {
                size_t chop = (line.size() - rpos) - 1;
                if (chop > 0) {
                  i -= chop;
                  line.erase(rpos + 1, chop);
                }
              }

              rpos = line.find_last_not_of(' ');

              if (rpos != std::string::npos) {
                size_t chop = (line.size() - rpos) - 1;
                if (chop > 0) {
                  line.erase(rpos + 1, chop);
                }
              }

              if (blankPending) {
                if (trace) {
                  output += "[PEND]";
                }
                if (spyDepth == lastSpyDepth) {
                  output += prefix;
                }
                output += '\n';
                blankPending = false;
              }

              if (spyDepth > 0) {
                output += prefix;
                lastSpyDepth = spyDepth;
              }

              output += line;
              if (trace) {
                output += "[WRAP]";
              }
              output += "\n";
              line.erase();
              continuation = false;

              column = 1;
              allWS = (line.find_first_not_of(' ') == std::string::npos);
              eatBlanks = true;
            }
        }
        break;

      case tsEscape:
        switch (ch) {
          case leAck:
            if (spyDepth == 0) {
              if (trace) {
                output += "[ACK]";
              }
              if (inState == inAckWait) {
                gettimeofday(&unthrottle, NULL);
                if (unthrottle.tv_usec < 500000) {
                  unthrottle.tv_usec += 500000;
                } else {
                  unthrottle.tv_sec += 1;
                  unthrottle.tv_usec -= 500000;
                }
                inState = inThrottle;
              }
            }
            state = tsData;
            break;

          case leSpy:
            state = tsSpyDepth;
            break;

          case leTrace:
            state = tsTrace;
            break;

          default:
            state = tsData;
        }
        break;

      case tsSpyDepth:
        if (trace) {
          sprintf(tbuf, "[S=%u]", ch);
          output += tbuf;
        }

        prefix = "";
        lastSpyDepth = spyDepth;
        spyDepth = ch;

        if (spyDepth > 0) {
          for (int j = 0; j < spyDepth; j++) {
            prefix += '/';
          }
          prefix += ' ';
        }

        state = tsData;
        break;

      case tsTrace:
        if (spyDepth == 0) {
          wasTrace = trace;
          trace = (ch != '-');

          if (trace || wasTrace) {
            sprintf(tbuf, "[T=%d]", trace);
            output += tbuf;
          }
        }
        state = tsData;
        break;
    }
  }

  if (!line.empty()) {
    if (blankPending) {
      if (spyDepth == lastSpyDepth) {
        output += prefix;
      }
      output += '\n';
      blankPending = false;
    }
    if (spyDepth > 0) {
      output += prefix;
      lastSpyDepth = spyDepth;
    }
    output += line;
    continuation = true;
  }

  if (!output.empty()) {
    writeOutput(output.data(), output.size());
  }
}

static void
readInput(int pipe) {
  (void) pipe;

  static unsigned char line[256];
  static size_t len = 0;

  unsigned char buf[BUFSIZ];
  const ssize_t nbytes = read(STDIN_FILENO, buf, sizeof(buf));

  if (nbytes <= 0) {
    if (nbytes == -1) {
      fatal("read");
    }
    exit(EXIT_SUCCESS);
  }

  iblock_t *b;

  for (ssize_t i = 0; i < nbytes; i++) {
    const unsigned char ch = buf[i];

    switch (ch) {
      case 0x00:
      case 0x01: // Ctrl-A
      case 0x02: // Ctrl-B
      case 0x03: // Ctrl-C
      case 0x04: // Ctrl-D
      case 0x05: // Ctrl-E
      case 0x06: // Ctrl-F
      case 0x07: // Ctrl-G
      case 0x08: // Ctrl-H
      case 0x09: // Ctrl-I
        break;

      case 0x0A: // Ctrl-J
        while (len > 0 && line[len - 1] == ' ') {
          len--;
        }
        line[len++] = '\n';

        b = static_cast<iblock_t *>(malloc(sizeof(iblock_t) + len));

        if (b == NULL) {
          fatal("malloc");
        }

        b->next = &iqTail;
        b->prev = iqTail.prev;
        b->len = len;
        memcpy(b->data, line, len);

        b->prev->next = b;
        iqTail.prev = b;

        len = 0;
        break;

      case 0x0B: // Ctrl-K
      case 0x0C: // Ctrl-L
      case 0x0D: // Ctrl-M
      case 0x0E: // Ctrl-N
      case 0x0F: // Ctrl-O
      case 0x10: // Ctrl-P
      case 0x11: // Ctrl-Q
        break;

      case 0x12: // Ctrl-R
        if (len > 0) {
          output(line, len);
        }
        break;

      case 0x13: // Ctrl-S
      case 0x14: // Ctrl-T
      case 0x15: // Ctrl-U
      case 0x16: // Ctrl-V
      case 0x17: // Ctrl-W
      case 0x18: // Ctrl-X
      case 0x19: // Ctrl-Y
      case 0x1A: // Ctrl-Z
      case 0x1B:
      case 0x1C:
      case 0x1D:
      case 0x1E:
      case 0x1F:
        break;

      case 0x7F:
        break;

      default:
        line[len] = ch;

        if (len < sizeof(line)) {
          len++;
        }
    }
  }
}

static void
sigWinch(int) {
  signal(SIGWINCH, sigWinch);
  winch = true;
}

static void
unavailable() {
  static const char message[] =
    "\n\nFederation is temporarily unavailable. Please try again later.\n\n";

  sleep(2);

  fputs(message, stdout);
  fflush(stdout);

  sleep(3);

  exit(EXIT_SUCCESS);
}

static void
warning(const char *s) {
  const int errnum = errno;

  perror(s);
  syslog(LOG_ERR, "%s: %s", s, strerror(errnum));
}

static void
writeOutput(const char *buf, size_t count) {
  if (oqHead.next == &oqTail) {
    ssize_t sent = write(STDOUT_FILENO, buf, count);

    if (sent == -1) {
      if (errno != EAGAIN) {
        fatal("write");
      }
      sent = 0;
    }

    buf += sent;
    count -= sent;
  }

  if (count == 0) {
    return;
  }

  oblock_t *b = static_cast<oblock_t *>(malloc(sizeof(oblock_t) + count));

  if (b == NULL) {
    fatal("malloc");
  }

  b->next = &oqTail;
  b->prev = oqTail.prev;
  b->sent = 0;
  b->len = count;
  memcpy(b->data, buf, count);

  b->prev->next = b;
  oqTail.prev = b;
}
